Add filtering and sorting support to JSON Store REST service #23 #25

Nikolay-S-Nikolov opened the pull request:
## Changes
- Added `sortBy` query parameter support for sorting by field (asc/desc)
- Added `where` query parameter support for filtering by field/value pairs
- Updated README with documentation and usage examples

## Usage
GET `/jsonstore/games?sortBy=_createdOn%20desc`
GET` /jsonstore/comments?where=gameId%3D%22{gameId}%22`

Closes #23
Closes #25

Nikolay-S-Nikolov committed:
Add sorting and filtering support to JSON Store GET responses

Nikolay-S-Nikolov committed:
add sorting and filtering documentation to JSONSTORE.md

viktorpts reviewed (changes requested):
This service was never intended to support more advanced operations, because we have the Collections service for that, but it's a welcome addition. Could we get some error handling, like we have in `common/crud.js`? If the query fails, the client will receive a generic error, or the server will outright crash.

Nikolay-S-Nikolov committed:
Add error handling in request handler and JSON store

Nikolay-S-Nikolov commented:
Added error handling for invalid query parameters (where/sortBy) to prevent server crashes.
Added validation for query strings in parseRequest() and returns RequestError instead of 500.

viktorpts commented on src/requestHandler.js:
```diff
@@ -104,12 +103,20 @@ async function parseRequest(req) {
+            .map((x) => {
+                const parts = x.split('=');
+                return parts
+            })
```
Perhaps you had more code going on here, but removed it before committing? In either case, this can be a single-line arrow function, as it was originally.

Nikolay-S-Nikolov replied:
Yes, you’re right. I had expanded it while debugging to provide a more precise error.

viktorpts commented on src/requestHandler.js:
```diff
@@ -104,12 +103,20 @@ async function parseRequest(req) {
+            .reduce((p, [k, v]) => Object.assign(p, { [k]: decodeURIComponent(v.replace(/\+/g, " ")) }), {});
+    } catch (err) {
+        console.error(err);
+        throw new RequestError(`Invalid query ${queryString}`);
```
Test how this will look if you append `err.message` to the response, since the client will not be able to see the reason their query failed otherwise. Granted, they should be looking at the server's output in their terminal, but we can make it easier to diagnose.

Nikolay-S-Nikolov committed:
Append err.message to the response
